Add protocol check for empty address in Service Entry

thehackercat opened the pull request:
Signed-off-by: LexusLee <lexuscyborg103@gmail.com>

**Please provide a description of this PR:**
Considering empty protocol might remove PassthroughCluster routing, add protocol validation when address is unset in ServiceEntry

![image](https://user-images.githubusercontent.com/10649416/174220900-9f029a98-bd36-4ea0-b30e-ff659db6d524.png)

Issue related: 
- https://github.com/istio/istio/issues/27990

we already have an analyzer for this case, I think we should also add a validation.

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes
Please check any characteristics that apply to this pull request.
[x] Does not have any changes that may affect Istio users.

istio-testing commented:
Hi @thehackercat. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

howardjohn reviewed (changes requested):
Making something that is valid invalid is a backwards compatible API change. We can add a warning potentially - although if we do it's probably more than just unset protocol (TCP would be the same?)

thehackercat commented:
> Making something that is valid invalid is a backwards compatible API change. We can add a warning potentially - although if we do it's probably more than just unset protocol (TCP would be the same?)

Thanks for the quick response, I've changed to warning level and added TCP protocol validation.

thehackercat commented:
> Can you add a release note: [https://github.com/istio/istio/tree/master/releasenotes](https://github.com/istio/istio/tree/master/releasenotes?rgh-link-date=2022-06-17T15%3A00%3A58Z)

done, PTAL

hzxuzhonghu commented:
/ok-to-test

istio-testing commented:
@thehackercat: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | ab948c10f5488ce48bcff520d52c7ec3ec1d7260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/39495/lint_istio/1538812695581560832) | true | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented on pkg/config/validation/validation.go:
```diff
@@ -3208,6 +3208,11 @@ var ValidateServiceEntry = registerValidateFunc("ValidateServiceEntry",
 				errs = appendValidation(errs, ValidatePort(int(port.TargetPort)))
 			}
+			if len(serviceEntry.Addresses) == 0 {
+				if port.Protocol == "" || port.Protocol == "TCP" {
```
Why TCP invalid?

thehackercat replied:
As explain in this comment https://github.com/istio/istio/pull/28664#issuecomment-725488225
When setting protocol to TCP with empty addresses, it makes all outgoing http traffic via port `xxxx` match `0.0.0.0_xxxx` tcp listener, which is unexpected.

hzxuzhonghu replied:
But 0.0.0.0_xxx would not happen when auto allocate ip for serviceentry.

hzxuzhonghu replied:
I agree with this solution since dns auto allocate is per proxy config

howardjohn commented on pkg/config/validation/validation.go:
```diff
@@ -3208,6 +3208,11 @@ var ValidateServiceEntry = registerValidateFunc("ValidateServiceEntry",
 			}
+			if len(serviceEntry.Addresses) == 0 {
+				if port.Protocol == "" || port.Protocol == "TCP" {
+					errs = appendValidation(errs, WrapWarning(fmt.Errorf("one or more protocol invalid in service entry with addresses empty")))
```
"invalid" is incorrect here. It is valid - it is just probably a bad idea. 

Maybe we can copy the similar text from the analyzer?

thehackercat replied:
it's done.

howardjohn commented on pkg/config/validation/validation.go:
```diff
@@ -3208,6 +3208,11 @@ var ValidateServiceEntry = registerValidateFunc("ValidateServiceEntry",
 			}
+			if len(serviceEntry.Addresses) == 0 {
+				if port.Protocol == "" || port.Protocol == "TCP" {
+					errs = appendValidation(errs, WrapWarning(fmt.Errorf("one or more protocol empty in service entry with addresses empty")))
```
```suggestion
					errs = appendValidation(errs, WrapWarning(fmt.Errorf("addresses are required for ports serving TCP (or unset) protocol")))
```

istio-testing commented:
@thehackercat: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-security-multicluster_istio | b15b9e493f235c66bb9510a56d1fe510f8c01a05 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/39495/integ-security-multicluster_istio/1546691507333894144) | true | `/test integ-security-multicluster_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

thehackercat commented:
/retest
